### PR TITLE
Minimal implementation of JSON-RPC server

### PR DESCRIPTION
**Pull request by r4f4ss**
This PR is a minimal implementation of the API-HTTP server, that will provide ```eth``` endpoints as [specified](https://github.com/ethereum/portal-network-specs?tab=readme-ov-file#the-json-rpc-api).

This implementation uses as much as possible functions, routines, and structures from go-ethereum code like ```node.newHTTPServer```, ```rpcEndpointConfig```, etc. This design takes advantage of several verifications and configuration parameters to avoid exploits and attacks already known and solved in go-etherum.

The use of go-ethereum implementations also allows to take advantage of several control features already implemented like rpcprefix, corsdomain, 
batch-request-limit and etc.

Only a skeleton of the JSON-RPC server and an example API (```eth_clientVersion```) was implemented, to extend to another APIs using the module eth, one must add to the file ```<PATH>/shisui/portalnetwork/ethapi/api.go``` new exportable methods, for instance, the implementation of ```eth_blockNumber```:
```
func (p *PortalEthereumAPI) BlockNumber() int {
  // TODO - logic here
}

```

Example of shisui invocation with default HTTP server:
```
shisui --http
```

Example of request to the example eth method:
```
curl 127.0.0.1:8545/ -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_clientVersion","params":[],"id":1}'
```

Obs: The old method ```web3_clientVersion``` was kept in the module web3 and migrated to this new implementation, to avoid any incompatibilities with previous behavior.

Obs2: ```rpc.port``` default port was changed to 30304 to avoid conflicts with 8545 used historically by go-ethereum for api-http .

### CODE REVIEW

**Comment by GrapeBaBa** [comment]
@r4f4ss Thank you very much for contribution for shisui. The current Portal API Spec is [here](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/ethereum/portal-network-specs/assembled-spec/jsonrpc/openrpc.json&uiSchema%5BappBar%5D%5Bui:splitView%5D=false&uiSchema%5BappBar%5D%5Bui:input%5D=false&uiSchema%5BappBar%5D%5Bui:examplesDropdown%5D=false), we only implemented `discv5` and `portal history` namespace API right now and we didn't start `eth` namespace.
You are correct that we should resue go-ethereum impl so that we used `rpc.server` struct and register new `portal` namespace for exposing the API of shisui. I think the new flags you introduced is good, I will look deeply for that and give feedback. @fearlessfe Please also take a look.

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -1035,6 +1035,68 @@ Please note that --` + MetricsHTTPFlag.Name + ` must be set to start the server.
...
+		Value:    node.DefaultConfig.BatchRequestLimit,
+		Category: flags.PortalNetworkApi,
+	}
+
+	PortalBatchResponseMaxSize = &cli.IntFlag{
+		Name:     "rpc.batch-response-max-size",
+		Usage:    "Maximum number of bytes returned from a batched call",
+		Value:    node.DefaultConfig.BatchResponseMaxSize,
+		Category: flags.PortalNetworkApi,
+	}
```
Comment (fearlessfe) [inline cmd/utils/flags.go:1099]: Maybe we don't need to add new flags, we can just reuse the exist flags

File: node/portal_node.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		CorsAllowedOrigins: config.HTTPCors,
+		Vhosts:             config.HTTPVirtualHosts,
+		Modules:            []string{"eth", "web3"},
+		prefix:             config.HTTPPathPrefix,
+		rpcEndpointConfig:  rpcConfig,
+	}); err != nil {
+		return err
+	}
+
+	if err := rpcServer.start(); err != nil {
```
Comment (fearlessfe) [inline node/portal_node.go:40]: We also start a http server in the `main.go`, we should just start one http server, the `history` and `beacon` server shoud as an Api here

Comment (r4f4ss) [reply node/portal_node.go:40]: I believe we indeed should start two or more servers, this is the design of go-ethereum, for example. It allows to set authentication, and other configuratins for api-http server while discV5 should be public and have diferent configurations.

**Comment by r4f4ss** [comment]
Thank you guys a lot for the feedback!

File: node/portal_node.go (hunk 2 above)
Comment (fearlessfe) [reply node/portal_node.go:40]: We just reuse some code of go-ethereum, the design of portal network is different. Now, portal network only has `history`, `beacon`, `state`, `web3` http module, these module should be serve at the same port. Or hive tests will be failed

**Comment by r4f4ss** [comment]
According to this Portal Network [explanation](https://ethereum.org/en/developers/docs/networking-layer/portal-network/#how-does-portal-network-work) from Ethereum documentation:

> The core idea of the Portal Network is to take the best bits of the current networking stack by enabling information needed by light clients, such as historical data and the identity of the current head of the chain to be served through a lightweight DevP2P style peer-to-peer decentralized network using a [DHT(opens in a new tab)](https://en.wikipedia.org/wiki/Distributed_hash_table) (similar to Bittorrent).

the design choice of create new servers is wrong. Feedback also showed limitations of this implementation.

While some flags may be a good contribuition, it is needed further study to confirm it is really needed, also to determine the default parameters value like ```evmtimeout```.

**Pull request closed without merging**